beancount/parser/lexer.l

```diff
@@ -573,3 +577,27 @@ void build_lexer_error_from_exception()
                         "Internal error: No exception");
     }
 }
+
+int PyFile_Read(PyObject *file, char *buf, size_t max_size)
+{
+     PyObject* dest = NULL;
+     PyObject* read = NULL;
+     int ret = 0;
+
+     dest = PyMemoryView_FromMemory(buf, max_size, PyBUF_WRITE);
```

Comment:
Hmm, on every single block read, this will (a) allocate an object, and (b) copy the buffer into it.
Right now the lexer just uses yy_scan_string() to set the buffer (there's one copy overall).
Frankly I don't think it matters... but where does this save a copy? (From the CL description)
Do you mean that instead of copying the entire string once and then scanning, the copy will only occur block-by-block?

Also, do you think there's an easy way to save and reuse the memoryview instead of recreating and deleting on every call to yyinput?

Reply:
Indeed this implementation is not the most efficient I can think of. I implemented an optimized one but it did not make any difference in a simple benchmark. Creating the memoryview is probably very cheap (I have benchmarked this but not profiled it). What I thought would have been expensive is looking up the "readinto" method and build the argument tuple. However removing those steps did not make any difference. I'll profile it and see.

It is a long time since I wrote this commit message and I am note sure anymore about saving a copy. With the current code:

1. the docstring is copied into a `char*` by the Python function argument unpacking,
2. that is copied into the lex buffer by `yy_scan_string()`,

thus the data is copied twice. With this code:

1. we docstring is wrapped into an `io.BytesIO` object (no copy),
2. the `io.BytesIO` object is copied into the lex buffer,

thus the data is copied once. However, the docstrings are unicode and not utf8, thus there they need to be encoded in utf8 before being wrapped into the `io.BytesIO`. I don't know if Python is smart enough to elide the copy if the strings are ASCII (like most of the test data). What would surely save one copy would be have the docstrings actually being bytes objects (place a `b` in front of the docstrings used as input data for the tests).